### ISSUE TITLE
Issue with nanotime min/which.min functions

### ISSUE THREAD

**Issue by statquant** [eddelbuettel/nanotime#53]
Hello, there is an ungoing issue with `integer64` that I think affect `nanotime`
```
min(nanotime(as.integer64(7300)), nanotime(as.integer64(7400)))
[1] "1970-01-01T01:00:00.000007400"
which.min(nanotime(c(as.integer64(7400)), as.integer64(7300)))
[1] 1
```
It looks very much like the same thing and referred to in [this SO question](https://stackoverflow.com/questions/58916594/why-does-min200-300-returns-300-for-integer64-class?noredirect=1#comment104094803_58916594)

**Comment by eddelbuettel** [comment]
Yes, thanks, we saw it too.

**Comment by statquant** [comment]
Hello, 
FYI a colleague of mine got bit by this bug today through `nanoduration` came to this repo and did not find this issue because it was closed.
Given you reference `bit64` in the `README` already, it might be useful for users to know than they should not use basic min/max functions on `nanotime/duration`.
```
> min(as.nanoduration(300), as.nanoduration(100))   
[1] 00:00:00.000_000_100                                     
> min(as.nanoduration(100), as.nanoduration(300))   
[1] 00:00:00.000_000_300                                     

```
Regards

**Comment by lsilvest** [comment]
Ouch, missed this. Thanks for reporting. Will provide a fix that is independent of `bit64`.

**Comment by statquant** [comment]
If you can get the bit64 maintainer to update the package I have the fix
for it...
Regards

**Comment by lsilvest** [comment]
I've checked that @truecluster already has a fix for that in `bit64` in his [latest version](https://github.com/truecluster/bit64). So let's see how close he is to being able to get his latest version on CRAN... If he's a while away, I'll define these functions for `nanotime` and `nanoduration`.

**Comment by statquant** [comment]
Now I feel stupid because I did not know about @truecluster and this github repo, it is not mentioned on the cran package description.

**Comment by truecluster** [comment]
I made several attempts to submit bit/bit64/ff to CRAN (after they passed all --as-cran checks on my machines) and they were repeatedly rejected. I hope this goes through soon - although my patience is quickly approaching zero.

**Comment by eddelbuettel** [comment]
I find the RHub builders valuable. It allowed to "develop" a little for macOS and solaris.

**Comment by truecluster** [comment]
I had mutual dependencies *and* small interface changes between bit and ff which leads to Errors in sequential package checking.  Plus newly invented checks, plus false Alarm, endless fun. Can RHub install multiple packages before checking them?

**Comment by eddelbuettel** [comment]
No, there you're fried.  I find Docker convenient there for my CI runs as I am in control of what goes in and how and when.

**Comment by lsilvest** [comment]
@truecluster: if you think I can be of help with troubleshooting CRAN's rejections of bit/bit64/ff let me know. Happy to try to help. Is it specifically MacOS and Solaris, or other stuff too?

**Comment by eddelbuettel** [comment]
I think @truecluster managed to get the set updated at CRAN.  Any chance you work your magic to fix min/max as discussed?

**Comment by lsilvest** [comment]
`bit64` on the CRAN web page still seems to indicate the 2017 version 0.9-7. Am I looking at outdated information? Also, I believe that once `bit64` is updated we should see correct min/max in `nanotime` without changes to our code.

**Comment by eddelbuettel** [comment]
Gosh, I am an idiot. I was in a (pre-Rcpp online tutorial for useR!) rush and misread. Sorry about the noise.

@truecluster Any update from your end?

**Comment by truecluster** [comment]
F.y.i.: nanotime is at risk.CRAN maintainers have received updated packages bit/bit64/ff for quite a while. They ist tortured me with multiple check complaints, mainly because they refuse to check the packages together. I removed as much of the problems as possible, now it depends on the willingness of CRAN but I don't see them act. I fear they prefer to orphan they old packages over uploading the new ones (into which I invested weeks). My patience is exhausted. At the same time Ripley warned me to kick bit64 out if I don't submit certain problems (that are fixed in the new version). If that happens I am gone.
To make a long story short: if you can influence the maintainers: give it a try. 
BestJens





Von: Leonardo Silvestri <notifications@github.com>
Gesendet: Dienstag, 14. Juli 2020, 16:54
An: eddelbuettel/nanotime
Cc: Jens; Mention
Betreff: Re: [eddelbuettel/nanotime] Issue with nanotime min/which.min functions (#53)






bit64 on the CRAN web page still seems to indicate the 2017 version 0.9-7. Am I looking at outdated information? Also, I believe that once bit64 is updated we should see correct min/max in nanotime without changes to our code.

**Comment by eddelbuettel** [comment]
I can't influence them either, and I am pretty `sample(c("mad", "tired", "frustrated", "worn out"), 1)`  too -- they gave a new (a somewhat involved, for work) package a two week run-around as a 'new' package, then tossed it within 24 hours because of "you-know-who".  Resubmitted 13 days ago, just sits there.  *Extremely frustrating* esp as I thought I had a decent working relationship with them.

**Comment by truecluster** [comment]
Then maybe their non-responding is due to User!2020

**Comment by eddelbuettel** [comment]
Cute, but no.  

I run [CRANberries](https://dirk.eddelbuettel.com/cranberries/) which chronciles a bazillion updates to packages as well as another bazillion of new packages.   Of course they are busy.  But a 13 days (and counting) non-response is, quite frankly, more than just borderline passive-aggressive nonsense.

**Comment by eddelbuettel** [comment]
So `bit64` is [now at 0.9.7-1.1](http://dirk.eddelbuettel.com/cranberries/2020/07/15#bit64_0.9-7.1) which looks like "you-know-who" updated without you. Correct?

**Comment by truecluster** [comment]
Correct, that reduces the risk that CRAN kicks out bit64 due to LTO errors in ther next weeks. But that increases the risk that I stop maintaining bit64. I keep you in the loop.


 

Gesendet: Mittwoch, 15. Juli 2020 um 14:45 Uhr
Von: "Dirk Eddelbuettel" <notifications@github.com>
An: "eddelbuettel/nanotime" <nanotime@noreply.github.com>
Cc: "Jens" <jens.oehlschlaegel@truecluster.com>, "Mention" <mention@noreply.github.com>
Betreff: Re: [eddelbuettel/nanotime] Issue with nanotime min/which.min functions (#53)


 

So bit64 is now at 0.9.7-1.1 which looks like "you-know-who" updated without you. Correct?

**Comment by eddelbuettel** [comment]
This should now be fixed in version 0.3.0 which is on CRAN (and will need a touchup for some build issues):

```r
R> min(nanotime(as.integer64(7300)), nanotime(as.integer64(7400)))
[1] 1970-01-01T00:00:00.000007300+00:00
R> which.min(nanotime(c(as.integer64(7400)), as.integer64(7300)))
[1] 1
R>
```
